src/optimizer/binder.py

```diff
@@ -0,0 +1,74 @@
+# coding=utf-8
+# Copyright 2018-2020 EVA
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import itertools
+
+from src.optimizer.rules.pattern import Pattern
+from src.optimizer.group_expression import GroupExpression
+from src.optimizer.memo import Memo
+from src.optimizer.operators import Operator, OperatorType
+
+
+class Binder:
+    def __init__(self, grp_expr: GroupExpression, pattern: Pattern, memo: Memo):
+        self._grp_expr = grp_expr
+        self._pattern = pattern
+        self._memo = memo
+
+    @staticmethod
+    def _grp_binder(id: int, pattern: Pattern, memo: Memo):
+        grp = memo.groups[id]
+        grp_exprs = grp.logical_exprs
+
+        for expr in grp.logical_exprs:
+            yield from Binder._binder(expr, pattern, memo)
```

Comment:
I think I was confused here. 

To my understanding, binder returns all matches for a group expression. Am I correct?

Also what does _grp_binder do vs _binder?

Reply:
Yes. You are right. Since logical_exprs only has one expression in the current implementation, so there is no difference.